driver/health/event_manager_default.go

```diff
@@ -0,0 +1,63 @@
+package health
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/ory/x/healthx"
+	"github.com/pkg/errors"
+)
+
+type DefaultHealthEventManager struct {
+	evtChan                chan interface{}
+	listeners              []Readiness
+	listenerEventTypeCache map[reflect.Type]Readiness
+}
+
+func NewDefaultHealthEventManager(listeners ...Readiness) (*DefaultHealthEventManager, error) {
+	var listenerEventTypeCache = make(map[reflect.Type]Readiness)
+	for _, listener := range listeners {
+		for _, evtType := range listener.EventTypes() {
+			evtTypeVal := reflect.TypeOf(evtType)
+			if _, ok := listenerEventTypeCache[evtTypeVal]; ok {
+				return nil, errors.WithStack(ErrEventTypeAlreadyRegistered)
+			}
+			listenerEventTypeCache[reflect.TypeOf(evtType)] = listener
+		}
+	}
```

Comment:
I've never actually used reflection so I don't really understand what's going on here. Maybe you could explain the idea behind this? :)

Reply:
The idea behind those few lines is to generate a map that associate an event type (aka a struct type) to a listener. 
The general idea is that `reflect.TypeOf` returns a `reflect.Type` that can be compared to another one to ensure that two different variables have the same type.

So for that, I iterate through each listeners I receive in argument, for each listener I call the `EventTypes` method which returns me a slice of interface that contains an instance of each event struct the probe needs to watch.
In the case of my rule readiness probe, I need to watch `readiness.RuleLoadedEvent` so the `EventTypes` function looks like something like this: 
```go
return []interface{}{&RuleLoadedEvent}
```
Then for each event types, I compute the `reflect.Type` value that I store in a map with the listener.

So now, when I receive an event to dispatch I simply have to compute its "type" using reflection, to check if I have something in my map, and if it's the case, to call the `EventsReceiver` function.

The reason of my choice for reflection is in my message below: https://github.com/ory/oathkeeper/pull/674#issuecomment-805292010